sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java

```diff
@@ -95,64 +95,88 @@ public Void apply(final Row row)
     };
   }
 
-  public static void executeSelect(
-      final DruidTable druidTable,
+  public void accumulate(
+      final DataSource dataSource,
+      final RowSignature sourceRowSignature,
+      final DruidQueryBuilder queryBuilder,
+      final Function<Row, Void> sink
+  )
+  {
+    if (dataSource instanceof QueryDataSource) {
+      final GroupByQuery outerQuery = queryBuilder.toGroupByQuery(dataSource, sourceRowSignature);
+      if (outerQuery == null) {
+        // Bug in the planner rules. They shouldn't allow this to happen.
+        throw new IllegalStateException("Can't use QueryDataSource without an outer groupBy query!");
+      }
+
+      executeGroupBy(queryBuilder, outerQuery, sink);
+      return;
+    }
+
+    final TimeseriesQuery timeseriesQuery = queryBuilder.toTimeseriesQuery(dataSource, sourceRowSignature);
+    if (timeseriesQuery != null) {
+      executeTimeseries(queryBuilder, timeseriesQuery, sink);
```

Comment:
I thought of having just one `toQuery` method, but the problem with that is then when we want to execute the query, we need to pair it with the correct execution strategy for that query (select needs to issue multiple queries for pagination, all query types have different result formats, etc). So that's why `accumulate` checks each possible query type individually.

I don't have a plan for changing this but I am open to change. Maybe DruidQueryBuilder should return both a Query and some kind of execution strategy object for it?